### ISSUE TITLE
IE11 ETag issue

### ISSUE THREAD

**Issue by SamPotts** [TTLabs/EvaporateJS#130]
We had a few complaints from IE users that uploads weren't working. We're using the awesome S3 caching feature to save uploading the same file multiple times. I had a look into the code as to what would be causing it and it looks like [line 600](https://github.com/TTLabs/EvaporateJS/blob/master/evaporate.js#L600):
```javascript
me.name = awsKey;
```
This was causing us issues because the check for the file was failing, and the attempt to kick off a new upload was working (as it should) but it was attempting to upload to a different path for some reason. I'm not sure why this is different to the `onErr` handler above it as that doesn't seem to set `me.name`. 

Anyhow, this aside, why was it getting this far when the file existed on S3? I done some more digging. It looks like the ETag property that IE returns in the XHR has an added `-1` on the end. For example, adding this into the method:

```javascript
console.log(eTag, me.eTag);
```
Would return in IE:
```
"b0a64cb88558c72a4b09778aeed5e017-1" "b0a64cb88558c72a4b09778aeed5e017"
```

My suggestions would be to remove line 600 to match the onErr handler and to split the ETag by "-" and use the first part. For example:

```javascript
var eTag = xhr.getResponseHeader('Etag').split("-")[0];
```

**Comment by SamPotts** [comment]
On further investigation it looks like the problem is related to the dash missing in `me.eTag` rather than the other way around

**Comment by SamPotts** [comment]
Looks like it's an issue with XML and IE:
https://social.msdn.microsoft.com/Forums/en-US/061721c1-68a5-4f38-83ec-d7e4b596cc1d/problem-parsing-xml-with-a-hyphen-in-ie11?forum=iewebdevelopment

**Comment by bikeath1337** [comment]
Ugh. Good old IE. Anyway, at least there's a way to address this now but I hate polluting the source with workarounds for something that should be fundamentally deterministic by now: XML parsing.

**Comment by SamPotts** [comment]
OK I think I've found a fix. I believe the issue lies with the `nodeValue` method. Because it fetches the first child node, which may be a text node and then returns that node value, and IE seems to treat a hyphen as a seperate text node, it would only return the first part of the ETag in this case.

```javascript
function nodeValue(parent, nodeName) {
    return parent.getElementsByTagName(nodeName)[0].childNodes[0].nodeValue;
}
```

This can be fixed by using `textContent` instead:

```javascript
function nodeValue(parent, nodeName) {
    return parent.getElementsByTagName(nodeName)[0].textContent;
}
```

Source: http://stackoverflow.com/questions/22039982/ie10-11-cdata-with-hyphens-parsed-wrong

So to sum up, deleting line 600 (to fix the fallback) and change the nodeValue method as above seems to fix IE without any negative effect in other browsers.

**Comment by SamPotts** [comment]
Yep, good old IE. It needs to go away already :+1:

**Comment by bikeath1337** [comment]
Thank you. Can you just submit the PR with the handful of suggested changes, Sam?

**Comment by SamPotts** [comment]
Sure thing. Will do now.

**Comment by tomsaffell** [comment]
@SamPotts good catch! 

But if they fixed IE, what we _do_ all do ? ;)

**Comment by bikeath1337** [comment]
Hi, @tomsaffell , it turned out not to be an IE bug but a bug I created by using `.nodeValue` rather than `textContent`, so we don't have to worry about that, fortunately.

**Comment by tomsaffell** [comment]
oh well, next time! (actually, in fairness to MSFT, IE 10+ seems pretty decent)

**Comment by bikeath1337** [comment]
I hope there is no next time for IE (!) and that it continues to be as decent as IE10. Glad it was an issue we could easily address to improve the project.

**Comment by SamPotts** [comment]
I pushed that change (to remove me.name being re-set) to production so lets see. Our UI tests passed though (with various scenarios) so I think it's good. I'll have a play over the weekend and report back. Was there any aspects in particular you'd like me to test?

**Comment by bikeath1337** [comment]
Thanks so much @SamPotts . I think you're right about the risk. When in the world is the eTag going to mismatch anyway? I think that is an edge case that can indeed be handed by the error case.

**Comment by bikeath1337** [comment]
@SamPotts I think we can close this.

**Comment by SamPotts** [comment]
Yeah it's been all good since that change 👍
